### DISCUSSION TITLE
Different middleware for different groups

### DISCUSSION THREAD

**Discussion by cemoktra** [tokio-rs/axum#292] [Q&A] [answered]
```
let app = Router::new()
    .route("/", get(root))
    .route("/some", get(some))
    .layer(auth)
    .route("/", post(root))
    .layer(auth)
    .layer(db)
    .route("/login", post(login))  // no layer
```

Is this valid code? The documentation just creates one group. What about multiple/different groups and combination with groups without layer.

**Comment by davidpdrsn** [comment, marked as answer]
While it is valid it probably doesn't do what you want. `layer` applies to all routes above it, meaning `auth` is applied twice to the first two routes.

You can do something like this instead:

```rust
let foo = Router::new()
    .route("/", get(root))
    .route("/some", get(some))
    .layer(auth);

let bar = Router::new()
    .route("/", post(root))
    .layer(auth)
    .layer(db);

let baz = Router::new()
    .route("/login", post(login));

let app = foo.or(bar).or(baz);
```

**Reply by cemoktra** [reply]
Thanks. Would be nice to have this in docs/examples

**Reply by davidpdrsn** [reply]
Good point! I'll add it right now :)

**Reply by hkaiser25** [reply]
Does the .or method still work on Router? I get this compiler error:
``no method named `or` found for struct `axum::Router` in the current scope
method not found in `Router<AppState>` ``
I'm using axum 0.7.4.

**Reply by jplatte** [reply]
`Router::or` was renamed to `merge` in v0.3.0, but the same principle still applies.
